### PR TITLE
Fixes #33268 - add power status dropdown to host details

### PR DESCRIPTION
**Pull request by amirfefer**
https://user-images.githubusercontent.com/11807069/130368314-a6c01276-1128-46d0-b08d-6022aa99493c.mov

### CODE REVIEW

File: webpack/assets/javascripts/react_app/components/HostDetails/DetailsCard/styles.scss [hunk 1]
```diff
@@ -6,3 +6,15 @@
...
+  color: var(--pf-global--success-color--100);
+}
+
+.power-off {
+  color: var(--pf-global--danger-color--100);
+}
+
+.power-na {
+  color: var(--pf-global--disable-color--200);
+}
```
Comment (MariaAga) [inline webpack/assets/javascripts/react_app/components/HostDetails/DetailsCard/styles.scss]: There should be an empty line at the end of the file

File: webpack/assets/javascripts/react_app/components/HostDetails/DetailsCard/PowerStatus/actions.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import { translate as __, sprintf } from '../../../../common/I18n';
+import { put } from '../../../../redux/API';
+import { POWER_REQURST_KEY, SUPPORTED_POWER_STATES } from './constants';
+
+export const changeHostPower = (state, hostID) =>
+  put({
+    key: POWER_REQURST_KEY,
+    params: { power_action: state },
+    url: `/api/hosts/${hostID}/power`,
+    errorToast: err => sprintf(__('an error occured: %s'), err),
```
Comment (MariaAga) [inline webpack/assets/javascripts/react_app/components/HostDetails/DetailsCard/PowerStatus/actions.js]:
``` suggestion
    errorToast: err => sprintf(__('an error occurred: %s'), err),
```

File: webpack/assets/javascripts/react_app/components/HostDetails/DetailsCard/PowerStatus/actions.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+        SUPPORTED_POWER_STATES[state]
+      ),
+    updateData: (prevState, { power }) => {
+      if (power)
+        return {
+          ...prevState,
+          state,
+          title: SUPPORTED_POWER_STATES[state],
+        };
+      return { ...prevState };
```
Comment (MariaAga) [inline webpack/assets/javascripts/react_app/components/HostDetails/DetailsCard/PowerStatus/actions.js]:
```suggestion
      return prevState;
```

File: webpack/assets/javascripts/react_app/components/HostDetails/DetailsCard/PowerStatus/PowerStatusIcon.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
+import PropTypes from 'prop-types';
+import React from 'react';
+import { PowerOffIcon } from '@patternfly/react-icons';
+import { Spinner } from '@patternfly/react-core';
+import { STATUS } from '../../../../constants';
+import '../styles.scss';
```
Comment (MariaAga) [inline webpack/assets/javascripts/react_app/components/HostDetails/DetailsCard/PowerStatus/PowerStatusIcon.js]: Won't it be loaded in `webpack/assets/javascripts/react_app/components/HostDetails/DetailsCard/index.js`?

File: webpack/assets/javascripts/react_app/components/HostDetails/DetailsCard/PowerStatus/PowerStatusDropDown.js [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+  const powerURL = foremanUrl(`/api/hosts/${hostID}/power`);
+  const {
+    response: { state: currentState, title, statusText },
+    status: responseStatus,
+  } = useAPI('get', powerURL, POWER_REQUEST_OPTIONS);
+
+  const dispatch = useDispatch();
+  const [isOpen, setOpen] = useState(false);
+  const changePowerHandler = targetState => {
+    // TODO: replace with confirmation service
```
Comment (MariaAga) [inline webpack/assets/javascripts/react_app/components/HostDetails/DetailsCard/PowerStatus/PowerStatusDropDown.js]: Can this be replaced?

File: webpack/assets/javascripts/react_app/components/HostDetails/DetailsCard/PowerStatus/PowerStatusDropDown.js [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+      >
+        {supportedPowerStates[state]}
+      </DropdownItem>
+    ));
+  };
+
+  const onDropdownSelect = event => setOpen(false);
+  const onToggle = open => setOpen(open);
+  return (
+    <Tooltip content={statusText || title}>
```
Comment (MariaAga) [inline webpack/assets/javascripts/react_app/components/HostDetails/DetailsCard/PowerStatus/PowerStatusDropDown.js:63]: When there is an error in the api call `statusText` and `title` are undefined. This causes an empty tooltip.
Should the dropdown be shown if there is an error?
Screenshot from 2021-09-07 20-36-28

**Review by MariaAga — DISMISSED** [review]
I don't have real hosts so I can't test it, left some code comments:

File: webpack/assets/javascripts/react_app/components/HostDetails/DetailsCard/styles.scss (hunk 1 above)
Comment (amirfefer) [reply webpack/assets/javascripts/react_app/components/HostDetails/DetailsCard/styles.scss]: should we add a rule or the css module plugin to our linter?

File: webpack/assets/javascripts/react_app/components/HostDetails/DetailsCard/PowerStatus/PowerStatusDropDown.js (hunk 6 above)
Comment (amirfefer) [reply webpack/assets/javascripts/react_app/components/HostDetails/DetailsCard/PowerStatus/PowerStatusDropDown.js:63]: if there is any error during fetching the API call, [statusText](https://github.com/theforeman/foreman/blob/develop/app/controllers/api/v2/hosts_controller.rb#L281) should contain the error, what error do you get from the server?

Comment (MariaAga) [reply webpack/assets/javascripts/react_app/components/HostDetails/DetailsCard/PowerStatus/PowerStatusDropDown.js:63]: I see, this was probably a fake api error. So no issue here

File: webpack/assets/javascripts/react_app/components/HostDetails/DetailsCard/PowerStatus/PowerStatusDropDown.js [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+import PowerStatusIcon from './PowerStatusIcon';
+import {
+  POWER_REQUEST_OPTIONS,
+  BASE_POWER_STATES,
+  SUPPORTED_POWER_STATES,
+} from './constants';
+import { changeHostPower } from './actions';
+import { openConfirmModal } from '../../../ConfirmModal';
+
+import '../styles.scss';
```
Comment (MariaAga) [inline webpack/assets/javascripts/react_app/components/HostDetails/DetailsCard/PowerStatus/PowerStatusDropDown.js:23]: Also, 
Won't it be loaded in webpack/assets/javascripts/react_app/components/HostDetails/DetailsCard/index.js?

**Review by MariaAga — COMMENTED** [review]
Great! can someone with real hosts take a look that it works? :)
@ezr-ondrej,@adamruzicka?

File: webpack/assets/javascripts/react_app/components/HostDetails/DetailsCard/PowerStatus/PowerStatusDropDown.js (hunk 7 above)
Comment (amirfefer) [reply webpack/assets/javascripts/react_app/components/HostDetails/DetailsCard/PowerStatus/PowerStatusDropDown.js:23]: Yes, though I didn't remove this import for keeping the power status component reuseable.

**Review by ezr-ondrej — APPROVED** [review]
Thanks @amirfefer ! Works as expected on libvirt and VmWare... I take it as good enough ;)

**Pull request merged**